cpp/inc/bond/protocol/simple_json_reader.h

```diff
@@ -179,15 +188,88 @@ class SimpleJsonReader
     friend typename boost::enable_if<is_map_container<X> >::type
     DeserializeMap(X&, BondDataType, const T&, SimpleJsonReader<Buffer>&);
 
-    SimpleJsonReader(const SimpleJsonReader& that, const char* name)
-        : _stream(that._stream),
-          _document(that._document),
-          _value(&(*that._value)[name])
-    {}
+    /// @brief Holds either an input stream XOR a pointer to some parent
+    /// StreamHolder.
+    class StreamHolder
+    {
+    public:
+        explicit StreamHolder(const Buffer& input)
+            : _stream(),
+              _parent()
+        {
+            // could use boost::in_place to construct this in the
+            // initializer list when the minimum version of Boost is 1.63
+            _stream.emplace(input);
+            BOOST_ASSERT(IsParent());
+        }
+
+        explicit StreamHolder(SimpleJsonReader& parent)
+            : _stream(),
+              // If we're creating a child of a child, use the "root" parent
+              // as our parent instead of having a bunch of intermediate
+              // children as parent.
+              _parent(parent._streamHolder.GetParent())
+        {
+            BOOST_ASSERT(_parent->IsParent());
+            BOOST_ASSERT(!IsParent());
+        }
+
+        // Intentionaly deep copy. Copies of SimpleJsonReader are expected
+        // to be deep copies, even if they're made from children.
+        StreamHolder(const StreamHolder& other)
+            : _stream(other.GetStream()),
+              _parent()
+        {
+            BOOST_ASSERT(IsParent());
+        }
+
+        #if defined(_MSC_VER) && _MSC_VER < 1900
+        // MSVC cannot = default rvalue ctor or move-assign operators
+        StreamHolder(StreamHolder&& other)
+            : _stream(std::move(other._stream)),
+              _parent(std::move(other._parent))
+        { }
+
+        StreamHolder& operator=(StreamHolder&& other)
+        {
+            _stream = std::move(other._stream);
+            _parent = std::move(other._parent);
+            return *this;
+        }
+        #else
+        StreamHolder(StreamHolder&&) = default;
+        StreamHolder& operator=(StreamHolder&&) = default ;
+        #endif
+
+        StreamHolder& operator=(const StreamHolder&) = default;
+
+        const detail::RapidJsonInputStream<Buffer>& GetStream() const
```

Comment:
Minor: you can rename to `Get`, the `Stream` part is redundant.

Reply:
✔️